contracts/OrderMixin.sol

```diff
@@ -119,76 +118,87 @@ abstract contract OrderMixin is IOrderMixin, EIP712, PredicateHelper {
+        return fillOrderTo(order, signature, interaction, input, threshold, target);
     }
 
     /**
      * @notice See {IOrderMixin-fillOrderTo}.
      */
     function fillOrderTo(
-        OrderLib.Order calldata order_,
+        OrderLib.Order calldata order,
         bytes calldata signature,
         bytes calldata interaction,
-        uint256 makingAmount,
-        uint256 takingAmount,
-        uint256 skipPermitAndThresholdAmount,
+        Input input,
+        uint256 threshold,
         address target
     ) public payable returns(uint256 actualMakingAmount, uint256 actualTakingAmount, bytes32 orderHash) {
-        if (target == address(0)) revert ZeroTargetIsForbidden();
-        orderHash = hashOrder(order_);
-
-        OrderLib.Order calldata order = order_; // Helps with "Stack too deep"
-        actualMakingAmount = makingAmount;
-        actualTakingAmount = takingAmount;
+        if (target == address(0)) {
+            target = msg.sender;
+        }
+        if (!order.constraints.isAllowedSender(msg.sender)) revert PrivateOrder();
+        if (order.constraints.isExpired()) revert OrderExpired();
+        if (!nonceEquals(order.maker.get(), order.constraints.series(), order.constraints.nonce())) revert WrongSeriesNonce();
 
+        orderHash = hashOrder(order);
         uint256 remainingMakingAmount = _remaining[orderHash];
         if (remainingMakingAmount == _ORDER_FILLED) revert RemainingAmountIsZero();
-        if (order.allowedSender.get() != address(0) && order.allowedSender.get() != msg.sender) revert PrivateOrder();
         if (remainingMakingAmount == _ORDER_DOES_NOT_EXIST) {
             // First fill: validate order and permit maker asset
             if (!ECDSA.recoverOrIsValidSignature(order.maker.get(), orderHash, signature)) revert BadSignature();
             remainingMakingAmount = order.makingAmount;
 
-            bytes calldata permit = order.permit();
-            if (skipPermitAndThresholdAmount & _SKIP_PERMIT_FLAG == 0 && permit.length >= 20) {
-                // proceed only if taker is willing to execute permit and its length is enough to store address
-                address token = address(bytes20(permit));
-                bytes calldata permitCalldata = permit[20:];
-                IERC20(token).safePermit(permitCalldata);
-                if (_remaining[orderHash] != _ORDER_DOES_NOT_EXIST) revert ReentrancyDetected();
+            if (!input.skipOrderPermit()) {
+                bytes calldata permit = order.permit();
+                if (permit.length >= 20) {
+                    // proceed only if taker is willing to execute permit and its length is enough to store address
+                    address token = address(bytes20(permit));
+                    bytes calldata permitCalldata = permit[20:];
+                    IERC20(token).safePermit(permitCalldata); // TODO: inline both arguments
```

Comment:
Why not TODO now?

Reply:
Can be refactored simultaneously in multiple places